bus/contractlocking.go

```diff
@@ -0,0 +1,205 @@
+	heldBy      uint64
+	queue       []*lockCandidate
+}
+
+type lockCandidate struct {
+	lockID       uint64
+	lockDuration time.Duration
+	c            chan struct{}
+}
+
+func newContractLocks() *contractLocks {
+	return &contractLocks{
+		locks: make(map[types.FileContractID]*contractLock),
+	}
+}
+
+func (l *contractLocks) lockForContractID(id types.FileContractID, create bool) *contractLock {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	lock, exists := l.locks[id]
+	if !exists && create {
+		c := make(chan struct{})
+		close(c)
+		lock = &contractLock{
+			lockChan:    c,
+			lockedUntil: time.Now().Add(time.Hour),
+		}
+		l.locks[id] = lock
+	}
+	return lock
+}
+
+// Acquire acquires a contract lock for the given id and provided duration. If
+// acquiring the lock doesn't finish before the context is closed,
+// ErrAcquireContractTimeout is returned. Upon success an identifier is returned
+// which can be used to release the lock before its lock duration has passed.
+// TODO: Extend this with some sort of priority. e.g. migrations would acquire a
+// lock with a low priority but contract maintenance would have a very high one
+// to avoid being starved by low prio tasks.
+func (l *contractLocks) Acquire(ctx context.Context, id types.FileContractID, d time.Duration) (uint64, error) {
+	lock := l.lockForContractID(id, true)
+
+	// Prepare a random lockID for ourselves.
+	ourLockID := frand.Uint64n(math.MaxUint64)
+
+	// Prepare a channel to indicate to other candidates whether we are
+	// still waiting to acquire the lock.
+	done := make(chan struct{})
+	defer close(done)
+
+	// Add ourselves to the queue of candidates for locking, grab the
+	// lockChan to be notified when the lock is released prematurely and
+	// prepare a timer to be notified when the current lock expires.
+	lock.mu.Lock()
+	lockChan := lock.lockChan
+	lock.queue = append(lock.queue, &lockCandidate{
+		lockID:       ourLockID,
+		lockDuration: d,
+		c:            done,
+	})
+	t := time.NewTimer(time.Until(lock.lockedUntil)) // prepare a timer for expiring locks
+	lock.mu.Unlock()
+
+	// Drain timer at the end if necessary.
+	drainTimer := func() {
+		if !t.Stop() {
+			select {
+			case <-t.C:
+			default:
+			}
+		}
+	}
+	defer drainTimer()
+
+	// tryAcquire is a helper function that tries to acquire a lock
+	// whenenver waiting threads are either woken by the lock timing out or
+	// the lock being released.
+	tryAcquire := func() (uint64, bool, error) {
+		lock.mu.Lock()
+
+		// Fetch the next candidate for locking. Drop any that
+		// have timed out in the meantime.
+		var nextCandidate *lockCandidate
+		for {
+			if len(lock.queue) == 0 {
+				panic("queue is empty - should never happen")
+			}
+			select {
+			case <-lock.queue[0].c:
+				lock.queue = lock.queue[1:]
+				continue
+			default:
+			}
+			nextCandidate = lock.queue[0]
+			break
+		}
+
+		// Check if it is safe to acquire the lock for the next
+		// candidate. That's the case if either lock.heldBy == 0 or we
```

Comment:
this is a little confusing, I would extract it into a method on the lock so it's more easily readable, maybe `IsAcquired` or `IsLocked` or `IsBusy`

Reply:
there's also a typo right, we're not beyond the locks expiry?